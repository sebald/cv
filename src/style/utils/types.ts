/**
 * The values will be treated like the following:
 * - number -> "px" value
 * - string -> "rem" value
 */
export type FontValue = number | string;

export type ScaleConfig = {
  fontSize: FontValue;
  htmlFontSize: number;
  ratio: number;
};

export type LineHeightConfig = {
  fontSize: FontValue;
  htmlFontSize: number;
  baseline: number;
};

export type TypographyConfig = ScaleConfig & LineHeightConfig;

export type WithClassName = {
  className?: string;
};
