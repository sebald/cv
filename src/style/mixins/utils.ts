import { CSSProperties } from 'react';
import { StyledMixin } from '../theme';

export const propOrEmpty = <P extends keyof CSSProperties>(
  name: P
): StyledMixin<Pick<CSSProperties, P>> => (props: Pick<CSSProperties, P>) => {
  const p = props[name];
  return p ? { [name]: p } : {};
};
