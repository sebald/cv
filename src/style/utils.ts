import { theme } from './theme';

const { htmlFontSize } = theme;

export const px = (value: number) => `${value}px`;

export const pxToRem = (pixel: number, base = htmlFontSize) =>
  `${parseFloat((pixel / base).toFixed(5))}rem`;

export const remToPx = (rem: string, base = htmlFontSize) =>
  parseFloat(rem) * base;

export const ensurePx = (value: number | string, base = htmlFontSize) =>
  typeof value === 'string' ? remToPx(value, base) : value;
