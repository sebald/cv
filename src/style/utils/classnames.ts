export const mergeClassNames = (...names: (string | undefined)[]) => ({
  className: names
    .filter(x => typeof x === 'string')
    .join(' ')
    .trim(),
});
