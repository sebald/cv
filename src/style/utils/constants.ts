export const htmlFontSize = 16;

export const ratio = {
  perfectFourth: 1.333,
  perfectFifth: 1.5,
  goldenRatio: 1.618,
  majorSixth: 1.667,
};
