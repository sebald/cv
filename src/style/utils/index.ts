export * from './classnames';
export * from './constants';
export * from './createMixinComponent';
export * from './unit';
export * from './typography';
