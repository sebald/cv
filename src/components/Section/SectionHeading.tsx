import * as React from 'react';

import { color, media, padding, styled } from '../../style';
import { Heading } from '../Typography';

const size = 3;

const SectionHeadingWrapper = Heading.extend`
  position: relative;
  text-align: center;

  &:before {
    content: '';
    display: block;

    height: ${size}px;
    width: 100%;

    background: ${color(palette => palette.grey['900'])};

    position: absolute;
    bottom: calc(50% - ${Math.abs(size / 2)}px);
    left: 0;

    z-index: 1;
  }

  ${media('small')`
    text-align: left;
  `};
`;
SectionHeadingWrapper.displayName = 'SeactionHeadingWrapper';

const SectionHeadingBackground = styled.span`
  position: relative;
  background: ${({ theme }) => theme.background.body};
  z-index: 5;

  ${padding(0, 2)} ${media('small')`
    text-align: left;
    ${padding(0, 2, 0, 0)}
  `};
`;
SectionHeadingBackground.displayName = 'SeactionHeadingBackground';

export const SectionHeading: React.SFC = ({ children }) => (
  <SectionHeadingWrapper>
    <SectionHeadingBackground>{children}</SectionHeadingBackground>
  </SectionHeadingWrapper>
);
