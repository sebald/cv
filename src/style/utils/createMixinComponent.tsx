import React from 'react';

import { styled } from '../styled-components';
import { StyledMixin } from '../theme';

import { mergeClassNames } from './classnames';
import { WithClassName } from './types';

/**
 * Idea stolen from:
 * https://github.com/jxnblk/grid-styled/tree/master/styled-space
 */
const StyledChildren: React.FC<WithClassName> = ({ className, children }) => {
  const styledChildren = React.Children.toArray(children).map((child) =>
    React.isValidElement<any>(child)
      ? React.cloneElement(
          child,
          mergeClassNames(child.props.className, className)
        )
      : child
  );

  return <>{styledChildren}</>;
};

export const createMixinComponent = <P extends {}>(
  ...mixins: StyledMixin<P>[]
) => styled(StyledChildren)([] as any, ...mixins);
