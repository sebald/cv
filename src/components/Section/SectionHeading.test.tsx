import * as React from 'react';

import { shallowWithTheme } from '../../test.utils';
import { SectionHeading } from './SectionHeading';

test('render correct', () => {
  const wrapper = shallowWithTheme(<SectionHeading />);
  expect(wrapper).toMatchSnapshot();
});

test('render children', () => {
  const wrapper = shallowWithTheme(<SectionHeading>child</SectionHeading>);
  expect(wrapper).toMatchSnapshot();
});
