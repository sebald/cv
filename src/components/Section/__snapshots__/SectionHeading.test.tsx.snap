// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render children 1`] = `
<SeactionHeadingWrapper>
  <SeactionHeadingBackground>
    child
  </SeactionHeadingBackground>
</SeactionHeadingWrapper>
`;

exports[`render correct 1`] = `
<SeactionHeadingWrapper>
  <SeactionHeadingBackground />
</SeactionHeadingWrapper>
`;
